dali/kernels/common/block_setup.h

```diff
@@ -0,0 +1,263 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_COMMON_BLOCK_SETUP_H_
+#define DALI_KERNELS_COMMON_BLOCK_SETUP_H_
+
+#include <cuda_runtime.h>
+#include <vector>
+#include <utility>
+#include "dali/kernels/kernel.h"
+#include "dali/core/geom/vec.h"
+
+namespace dali {
+namespace kernels {
+
+template <int ndim>
+struct BlockDesc {
+  int sample_idx;
+  ivec<ndim> start, end;
+};
+
+template <int n>
+DALI_HOST_DEV
+ivec<n> shape2vec(const TensorShape<n> &shape) {
+  ivec<n> ret;
+  for (int i = 0; i < n; i++)
+    ret[n-1-i] = shape[i];
+  return ret;
+}
+
+template <int skip, int n>
+DALI_HOST_DEV
+std::enable_if_t<(skip < 0), TensorShape<n>> skip_dim(const TensorShape<n> &shape) {
+  return shape;
+}
+
+template <int skip, int n>
+DALI_HOST_DEV
+std::enable_if_t<(skip >= 0), TensorShape<n-1>> skip_dim(const TensorShape<n> &shape) {
+  static_assert(skip < n, "The dimension to be skipped must not exceed input ndim");
+  return shape_cat(shape.template first<skip>(), shape.template last<n-skip-1>());
+}
+
+/// @brief A utility for calculating block layout for GPU kernels
+/// @tparam _ndim         number dimensions to take into account while calculating the layout
+/// @tparam _channel_dim  dimension in which channels are stored; channel dimension does not
+///                       participate in layout calculation \n
+///                       In cases where channel dimension can or should participate in layout
+///                       calaculation, do not specify channel dimenion and treat it as an
+///                       additional spatial dimension (e.g. for linear operations in CHW layout)\n
+///                       -1 indicates there are only spatial dimensions, all of which
+///                       participate in layout calculation.
+template <int _ndim, int _channel_dim>
+class BlockSetup {
+ public:
+  static constexpr int ndim = _ndim;
+  static constexpr int channel_dim = _channel_dim;
+  static constexpr int tensor_ndim = (channel_dim < 0 ? ndim : ndim + 1);
+  static_assert(channel_dim >= -1 && channel_dim <= ndim,
+    "Channel dimension must be in range [0..ndim] or -1 (no channel dim)");
+  using BlockDesc = kernels::BlockDesc<ndim>;
+
+  BlockSetup() {
+    for (int i = 0; i < ndim; i++)
+      default_block_size_[i] = (i < 2) ? 256 : 1;
+  }
+
+  void SetupBlocks(const TensorListShape<tensor_ndim> &output_shape,
+                   bool force_variable_size = false) {
+    blocks_.clear();
+    is_uniform_ = !force_variable_size && is_uniform(output_shape);
+
+    if (is_uniform_)
+      UniformSizeSetup(output_shape);
+    else
+      VariableSizeSetup(output_shape);
+  }
+
+  TensorListShape<tensor_ndim> GetOutputShape(
```

Comment:
What is its purpose? I cannot see any test or usage.

Reply:
Test added.